src/libtls/tls_server.c

```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
+	int *k;
+	key_type_t *v, lookup;
+	uint16_t sig_scheme;
+
+	ht = tls_get_supported_key_types(version_min, version_max);
+	if (!ht)
+	{
+		return NULL;
+	}
+	key_types = array_create(sizeof(key_type_t), 1);
+	enumerator = ht->create_enumerator(ht);
+	while (enumerator->enumerate(enumerator, &k, &v))
+	{
+		bio_reader_t *reader;
+
+		reader = bio_reader_create(hashsig);
+		while (reader->remaining(reader) &&
+			   reader->read_uint16(reader, &sig_scheme))
+		{
+			lookup = tls_signature_scheme_to_key_type(sig_scheme);
+			if (*v == lookup)
+			{
+				array_insert(key_types, ARRAY_TAIL, &lookup);
+				break;
+			}
+		}
+		reader->destroy(reader);
+	}
+	enumerator->destroy(enumerator);
+	ht->destroy_function(ht, (void*)free);
+	return key_types;
+}
+
 /**
  * Find a cipher suite and a server key
  */
 static bool select_suite_and_key(private_tls_server_t *this,
 								 tls_cipher_suite_t *suites, int count)
 {
+	array_t *key_types;
+	tls_version_t version_min, version_max;
 	private_key_t *key;
 	key_type_t type;
 
-	key = lib->credmgr->get_private(lib->credmgr, KEY_ANY, this->server,
-									this->server_auth);
+	version_min = this->tls->get_version_min(this->tls);
+	version_max = this->tls->get_version_max(this->tls);
+	key_types = create_common_key_types(this->hashsig, version_min, version_max);
+	if (!key_types || !array_count(key_types))
+	{
+		DBG1(DBG_TLS, "no common signature algorithms found");
+		array_destroy(key_types);
+		return FALSE;
+	}
+	while (array_remove(key_types, ARRAY_HEAD, &type))
+	{
+		key = lib->credmgr->get_private(lib->credmgr, type, this->server,
+										this->server_auth);
+		if (key)
+		{
+			break;
+		}
+	}
 	if (!key)
 	{
 		DBG1(DBG_TLS, "no usable TLS server certificate found for '%Y'",
 			 this->server);
+		array_destroy(key_types);
 		return FALSE;
 	}
 
-	if (this->tls->get_version_max(this->tls) >= TLS_1_3)
+	if (version_max >= TLS_1_3)
 	{
-		/* currently no support to derive key based on client supported
-		 * signature schemes */
 		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
 												  		count, KEY_ANY);
-		if (!this->suite)
-		{
-			DBG1(DBG_TLS, "received cipher suites unacceptable");
-			return FALSE;
-		}
 	}
 	else
 	{
-		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
-												  		count,
-												  		key->get_type(key));
-		if (!this->suite)
-		{	/* no match for this key, try to find another type */
-			if (key->get_type(key) == KEY_ECDSA)
-			{
-				type = KEY_RSA;
-			}
-			else
-			{
-				type = KEY_ECDSA;
-			}
-			key->destroy(key);
-
+		if (type)
+		{
 			this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
 															count, type);
-			if (!this->suite)
-			{
-				DBG1(DBG_TLS, "received cipher suites unacceptable");
-				return FALSE;
-			}
+		}
+		while (!this->suite && array_remove(key_types, ARRAY_HEAD, &type))
+		{	/* find a cipher suite that supports the key */
+			key->destroy(key);
```

Comment:
If we find no key for a type, this won't be defined anymore.
```suggestion
			DESTROY_IF(key);
```